pkg/agent/manager/cache/cache.go

```diff
@@ -2,357 +2,582 @@ package cache
+	set := allocSelectorSet(selectors...)
+	defer freeSelectorSet(set)
+
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.matchingEntries(set)
 }
 
-type Cache interface {
-	// FetchEntry gets the cache entry for the specified registration entry id
-	FetchEntry(entryId string) *Entry
-	// SetEntry puts a new cache entry for the entry's RegistrationEntry.
-	SetEntry(entry *Entry)
-	// DeleteEntry removes the cache entry for the specified RegistrationEntry if it exists,
-	// returns true if it removed some entry or false otherwise.
-	DeleteEntry(regEntry *common.RegistrationEntry) bool
-	// Entries returns all the in force cached entries.
-	Entries() []*Entry
-	// Registers and returns a Subscriber, and then sends latest WorkloadUpdate on its channel
-	Subscribe(selectors Selectors) Subscriber
-	// Set the bundles
-	SetBundles(map[string]*Bundle)
-	// Retrieve the bundle for the trust domain
-	Bundle() *Bundle
-	// SubscribeToBundleChanges returns a bundle stream. Each
-	// time bundles are updated, a new bundle mapping is streamed.
-	SubscribeToBundleChanges() *BundleStream
-	// FetchWorkloadUpdates gets the latest workload update for the selectors
-	FetchWorkloadUpdate(selectors Selectors) *WorkloadUpdate
+func (c *Cache) FetchWorkloadUpdate(selectors []*common.Selector) *WorkloadUpdate {
+	set := allocSelectorSet(selectors...)
+	defer freeSelectorSet(set)
 
-	// GetJWTSVID retrieves a cached JWT SVID based on the subject and
-	// intended audience.
-	GetJWTSVID(spiffeID string, audience []string) (*client.JWTSVID, bool)
-	// SetJWTSVID caches a JWT SVID based on the subject and intended audience.
-	SetJWTSVID(spiffeID string, audience []string, svid *client.JWTSVID)
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.buildWorkloadUpdate(set)
 }
 
-type cacheImpl struct {
-	// Map keyed by RegistrationEntry.EntryId holding Entry instances.
-	cache       map[string]*Entry
-	log         logrus.FieldLogger
-	m           sync.Mutex
-	subscribers *subscribers
-	trustDomain string
-	bundles     observer.Property
-	notifyMutex sync.Mutex
-	jwtSVIDS    map[string]*client.JWTSVID
+func (c *Cache) SubscribeToWorkloadUpdates(selectors []*common.Selector) Subscriber {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	sub := newSubscriber(c, selectors)
+	for s := range sub.set {
+		c.addSelectorIndexSub(s, sub)
+	}
+	c.notify(sub)
+	return sub
 }
 
-// New creates a new Cache.
-func New(log logrus.FieldLogger, trustDomain string, bundle *Bundle) *cacheImpl {
-	bundles := map[string]*Bundle{
-		trustDomain: bundle,
+func (c *Cache) Update(update *CacheUpdate, checkSVID func(*common.RegistrationEntry, *X509SVID)) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// remove bundles that no longer exist and update bundle with changes,
+	// populating a "changed" set that we can check when processing
+	// registration entries to know if they need to spawn a notification.
+	// never remove the bundle for the trust domain.
```

Comment:
Sure, I can split the comment. The "never remove the bundle for the trust domain" was a safety measure the previous cache implementation had in place to make sure that no matter what we had something to use to authorize the server (in case there was a bug where the server opted to not return the bundle). I'll try and clean up the language.